js/upload/src/utils.ts

```diff
@@ -65,69 +65,99 @@ function is_url(str: string): boolean {
 	}
 	if (is_url(path)) {
 		return path;
 	}
-	return root_url ? `/proxy=${root_url}file=${path}` : `${root}/file=${path}`;
+	return proxy_url
+		? `/proxy=${proxy_url}file=${path}`
+		: `${server_url}/file=${path}`;
 }
 
-export const blobToBase64 = (blob: File): Promise<string> => {
-	const reader = new FileReader();
-	reader.readAsDataURL(blob);
-	return new Promise((resolve) => {
-		reader.onloadend = (): void => {
-			resolve(reader.result as string);
-		};
-	});
-};
-
 export async function upload(
 	file_data: FileData[],
 	root: string,
-	upload_fn: typeof upload_files | undefined = upload_files
-): Promise<FileData[]> {
+	upload_fn: typeof upload_files = upload_files
+): Promise<(FileData | null)[] | null> {
 	let files = (Array.isArray(file_data) ? file_data : [file_data]).map(
 		(file_data) => file_data.blob!
 	);
 
-	await upload_fn(root, files).then(async (response) => {
-		if (response.error) {
-			(Array.isArray(file_data) ? file_data : [file_data]).forEach(
-				async (file_data, i) => {
-					file_data.data = await blobToBase64(file_data.blob!);
-					file_data.blob = undefined;
-				}
-			);
-		} else {
-			(Array.isArray(file_data) ? file_data : [file_data]).forEach((f, i) => {
-				if (response.files) {
-					f.orig_name = f.name;
-					f.name = response.files[i];
-					f.is_file = true;
-					f.blob = undefined;
-					normalise_file(f, root, null);
+	return await Promise.all(
+		await upload_fn(root, files).then(
+			async (response: { files?: string[]; error?: string }) => {
+				if (response.error) {
+					throw new Error(response.error);
+				} else {
+					if (response.files) {
+						return response.files.map((f, i) => {
+							const file = new FileData({ ...file_data[i], path: f });
+
+							return normalise_file(file, root, null);
+						});
+					}
+
+					return [];
 				}
-			});
-		}
-	});
-	return file_data;
+			}
+		)
+	);
 }
 
 export async function prepare_files(files: File[]): Promise<FileData[]> {
-	var all_file_data: FileData[] = [];
-	files.forEach((f, i) => {
-		all_file_data[i] = {
-			name: f.name,
-			size: f.size,
-			data: "",
-			blob: f
-		};
-	});
-	return all_file_data;
+	return files.map(
+		(f, i) =>
+			new FileData({
+				path: f.name,
+				orig_name: f.name,
+				blob: f,
+				size: f.size,
+				mime_type: f.type
+			})
+	);
+}
+
+export class FileData {
+	path: string;
+	url?: string;
+	orig_name: string;
+	size?: number;
+	blob?: File;
+	is_stream?: boolean;
+	mime_type?: string;
+	alt_text?: string;
+
+	constructor({
+		path,
+		url,
+		orig_name,
+		size,
+		blob,
+		is_stream,
+		mime_type,
+		alt_text
+	}: {
+		path: string;
+		url?: string;
+		orig_name: string;
+		size?: number;
+		blob?: File;
+		is_stream?: boolean;
+		mime_type?: string;
+		alt_text?: string;
+	}) {
+		this.path = path;
+		this.url = url;
+		this.orig_name = orig_name;
+		this.size = size;
+		this.blob = blob;
+		this.is_stream = is_stream;
+		this.mime_type = mime_type;
+		this.alt_text = alt_text;
+	}
```

Comment:
`FileData` is a class now, felt a little simpler and better parity with the backend.